fix 111933

cvvz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug

#### What this PR does / why we need it:

Fixes #111933, do not check the existence of mount point during reconstruction. If mount point does not exist, reconstruction will success and update reconstructed volume into ASW, which will be unmounted and cleaned up in the subsequent reconcile.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #111933

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

yangjunmyfm192085 commented:
/release-note-none

gnufied commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -375,6 +375,15 @@ func (c *csiMountMgr) TearDown() error {
+		return errors.New(log("Unmounter.TearDownAt failed to clean mount dir [%s]: %v", dir, err))
+	}
+
 	volID := c.volumeID
```
This does not seem correct, so if `NodeUnpublish` fails then subsequent calls to `NodeUnpublish`  won't even start because directory and json files has already been removed. 

I do not think fixing it here is the right solution but we need to think it from perspective of how cleanup should work if drivers are not *yet* available or volume is not yet mounted.  IMO - the fix has to be made in reconstruction code, not here.

cvvz replied:
Actually, csi mount directory will be removed anyway by housekeeping handler in [removeOrphanedPodVolumeDirs](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_volumes.go#L119), this function will even remove entire pod volume dir, which is `/var/lib/kubelet/pods/<POD_UID>/volumes`, the reason why the directory was not deleted sucessfully and left over is that there is a file called `vol_data.json`, and this function can only delete directory but not file. So here I put `removeMountDir` before calling out csi driver to make sure `vol_data.json` is cleaned up no matter `NodeUnpublish` sucess or not, in this way, the orphan pod volume will be cleaned up by housekeeping handler eventually.

dobsonj replied:
I agree with @gnufied , this doesn't seem like the right place to fix it.
What happens when the directory is still mounted? Removal of the directory would fail then but with this change it means `TearDownAt` would return an error before calling `NodeUnpublishVolume` to unmount the volume, and it would never get unmounted. So I think this breaks the main use case (unmount volume and then remove the mountpoint).

cvvz replied:
Actually, when the directory is still mounted, `removeMountDir` won't return error. It still should be unmounted by csi driver first. Every cleaning up action in kubelet should have considered the mount state in the first place.

Considering that csi driver is responsible for unmounting and cleaning the `mount` directory, maybe we could only remove `vol_data.json` here, which should be done by kubelet, rather than remove the mount dir before calling out `NodeUnpublishVolume`.

cvvz replied:
@gnufied @dobsonj I agree that we should not break the original logic of `TearDownAt`. And I update the PR. This time I put the deletion of the volume data file in kubelet housekeeping process. Could you PTAL?

dobsonj commented on pkg/kubelet/kubelet_volumes.go:
```diff
@@ -152,9 +153,14 @@ func (kl *Kubelet) removeOrphanedPodVolumeDirs(uid types.UID) []error {
+	// Fail if any regular files other than CSI volume data file are encountered.
 	podVolDir := kl.getPodVolumesDir(uid)
-	if err := removeall.RemoveDirsOneFilesystem(kl.mounter, podVolDir); err != nil {
+	if err := removeall.RemoveAllOneFilesystemCommon(kl.mounter, podVolDir, func(path string) error {
```
This change actually re-introduces a rather serious race condition where `vol_data.json` may be deleted by `removeOrphanedPodVolumeDirs` before `NodeUnstageVolume` has a chance to run, which causes unmount to fail and kubelet ends up not being able to clean the mountpoints on the node:
https://github.com/kubernetes/kubernetes/issues/101911
So this is not the right place to remove `vol_data.json`, it will result in unmounts failing in some scenarios.
You may want to take a look at the [volume reconstruction](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go) code in volumemanager. This is a tough bug in a complex part of the code, and the fix is not obvious to me, but when I saw it before [syncStates failed here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go#L506) because the /mount subdir did not exist. Then it calls `cleanupMounts` but that failed because the volume failed to be reconstructed and it couldn't create the unmounter at that point. If you're able to reproduce the problem, you can check if you're seeing the same failure scenario in kubelet.log.
IMO, we either need to be able to reconstruct the volume to unmount it in this scenario, or at least find some way to create the unmounter so that cleanupMounts can remove vol_data.json even when the /mount subdir does not exist. Sorry if this direction is still a little vague, but that's as far as I've been able to dig into this.

cvvz replied:
@dobsonj Thank you so much for the explanation. I am excited that there are still so many things I didn't know😄, and I will dig into it again to see if I can find a correct solution.

cvvz replied:
@dobsonj  @gnufied  Actually, not only the volume directory `/mount` will be left over, the `/globalmount` directory will be left over either. Because in this scenario, the volume manager did neither of `NodeUnstageVolume` nor `NodeUnpublishVolume`:

When reconstructing, the volume directory is not mountpoint, so volume manager tries to do cleanup by calling `NodeUnpublishVolume` and cleanup the `/mount` directory and `vol_data.json`, but the csi driver did not register yet, which makes cleanup fails. Actually, no matter it succeeds or fails, `NodeUnstageVolume` won't be called, so `/globalmount` will be left over.

I think there may be two solutions:
1. Let the reconstruction success no matter `/mount` is mountpoint or not. After adding the volume to ASW, the volume manager's reconcile process will do `NodeUnstageVolume` and `NodeUnpublishVolume` later. 
2. When the unmounter executes unmounting, clean up the directory and `vol_data.json` without calling `NodeUnpublishVolume` and `NodeUnstageVolume` if the directory is already not a mountpoint.

WDYT? Which one is better or any other idea?

cvvz committed:
reconstruct no matter mount point exists or not

cvvz commented:
@dobsonj  @gnufied  Hi, I've updated the PR. And the latest solution is not checking the existence of mount point during reconstruction. If mount point does not exist, reconstruction will success and update reconstructed volume into ASW, which will be unmounted and cleaned up in the subsequent reconcile. 

So, previously, why making reconstruction fail when mountpoint does not exist? I found the original issue https://github.com/kubernetes/kubernetes/issues/51982, it is to fix the problem that after rebooting the node, ASW and DSW will be the same after reconstruction and updating ASW succeed, thus reconcile won't remount the volume which is no longer mount point. However, this will never happen now since [we will not update ASW if the volume is in DSW](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go#L522-L532).

dims commented:
If you still need this PR then please rebase, if not, please close the PR
